text/2020-00-00-error-handling.md

```diff
@@ -0,0 +1,297 @@
+[src/storage](https://github.com/tikv/tikv/blob/dec3d008acb56ad0fb34eeeecca62aaa9091b994/src/storage/errors.rs#L174-L180).
+
+Due to having many error types, error types are often nested together to form
+chains. However, we rarely use error backtraces (most of our errors don't even
+support them) so this just adds code complexity and duplication (different error
+types with identical variants). In particular, when handling errors it leads to
+deep matching to find all instances of an error.
+
+## Detailed design
+
+Goals and constraints:
+
+* Use statically typed errors (c.f.,
+  [anyhow](https://github.com/dtolnay/anyhow), etc.) to eliminate many matching
+  errors (this constraint also precludes a single error type).
+* Use a single, modern, well-maintained library which is common in the wider
+  Rust ecosystem and supports `std::Error`.
+* Support boxing errors more ergonomically.
+* Reduce deep nesting of errors.
+* Reduce error boilerplate.
+* Reduce custom error handling code (e.g., our `box_try` macro).
+* Support incremental migration to the new error format.
+
+I propose using [thiserror](https://github.com/dtolnay/thiserror). We will
+modify thiserror to support *boxed* and *unwrapping* `From` implementations.
+Each error type will be migrated to a single enum (no inner enum). Large nested
+errors will be individually boxed (note that a boxed enum is the same size as an
+enum where every variant is boxed).
+
+### Examples
+
+Current definition (from
+[src/storage/txn/mod.rs](https://github.com/tikv/tikv/blob/dec3d008acb56ad0fb34eeeecca62aaa9091b994/src/storage/txn/mod.rs#L42-L94)
+):
+
+```rust
+quick_error! {
+    #[derive(Debug)]
+    pub enum ErrorInner {
+        Engine(err: crate::storage::kv::Error) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Mvcc(err: crate::storage::mvcc::Error) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        ProtoBuf(err: protobuf::error::ProtobufError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Io(err: IoError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        InvalidTxnTso {start_ts: TimeStamp, commit_ts: TimeStamp} {
+            description("Invalid transaction tso")
+            display("Invalid transaction tso with start_ts:{},commit_ts:{}",
+                        start_ts,
+                        commit_ts)
+        }
+        Other(err: Box<dyn error::Error + Sync + Send>) {
+            from()
+            cause(err.as_ref())
+            description(err.description())
+            display("{:?}", err)
+        }
+    }
+}
+```
+
+Definition with proposed changes:
+
+```rust
+#[derive(Debug, thiserror::Error)]
+#[error(unwrap)]
+pub enum Error {
+    #[error(transparent)]
+    Engine(#[from] crate::storage::kv::Error),
```

Comment:
seem we don't need to use ErrorInner any more and passing Error directly doesn't affect the performance?

Reply:
This error is already boxed so it doesn't need to be boxed here too. We still box some errors for performance.